jupyter/d2l-java/chapter_multilayer-perceptrons/underfit-overfit.ipynb

```diff
@@ -0,0 +1,981 @@
+    "\n",
+    "        Loss l2Loss = Loss.l2Loss();\n",
+    "        NDManager manager = NDManager.newBaseManager();\n",
+    "        LearningRateTracker lrt = LearningRateTracker.fixedLearningRate(0.01f);\n",
+    "        Optimizer sgd = Optimizer.sgd().setLearningRateTracker(lrt).build();\n",
+    "        DefaultTrainingConfig config = new DefaultTrainingConfig(l2Loss)\n",
+    "                .optOptimizer(sgd) // Optimizer (loss function)\n",
+    "                .addTrainingListeners(TrainingListener.Defaults.logging()); // Logging\n",
+    "\n",
+    "        Model model = Model.newInstance(\"mlp\");\n",
+    "        SequentialBlock net = new SequentialBlock();\n",
+    "        Linear linearBlock = Linear.builder().optBias(false).setOutChannels(1).build();\n",
+    "        net.add(linearBlock);\n",
+    "\n",
+    "        model.setBlock(net);\n",
+    "        Trainer trainer = model.newTrainer(config);\n",
+    "\n",
+    "        int batchSize = Math.min(10, (int) trainLabels.getShape().get(0));\n",
+    "\n",
+    "        ArrayDataset trainIter = loadArray(trainFeatures, trainLabels, batchSize, true);\n",
+    "        ArrayDataset testIter = loadArray(testFeatures, testLabels, batchSize, true);\n",
+    "\n",
+    "        trainer.initialize(new Shape(1, nDegree));\n",
+    "\n",
+    "        for (int epoch = 1; epoch <= numEpochs; epoch++) {\n",
+    "\n",
+    "            // Iterate over dataset\n",
+    "            for (Batch batch : trainer.iterateDataset(trainIter)) {\n",
+    "                // Update loss and evaulator\n",
+    "                EasyTrain.trainBatch(trainer, batch);\n",
+    "\n",
+    "                // Update parameters\n",
+    "                trainer.step();\n",
+    "\n",
+    "                batch.close();\n",
+    "            }\n",
+    "            // reset training and validation evaluators at end of epoch\n",
+    "\n",
+    "            for (Batch batch : trainer.iterateDataset(testIter)) {\n",
+    "                // Update loss and evaulator\n",
+    "                EasyTrain.validateBatch(trainer, batch);\n",
+    "\n",
+    "                batch.close();\n",
+    "            }\n",
+    "\n",
+    "            trainer.notifyListeners(listener -> listener.onEpoch(trainer));\n",
+    "            if (epoch % 50 == 0) {\n",
+    "                epochCount[epoch/50 - 1] = epoch;\n",
+    "                trainLoss[epoch/50 - 1] = trainer.getTrainingResult().getEvaluations().get(\"train_loss\");\n",
+    "                testLoss[epoch/50 - 1] = trainer.getTrainingResult().getEvaluations().get(\"validate_loss\");\n",
+    "            }\n",
+    "\n",
+    "        }\n",
+    "\n",
+    "        Block layer = model.getBlock();\n",
+    "        ParameterList params = layer.getParameters();\n",
+    "        weight = params.get(0).getValue().getArray();\n",
```

Comment:
Make function return you the weight

Reply:
I thought of showing the user the weights separately as the weights and the graph is the main point of the code snippet. If I keep it in the function, the user has the scroll down a lot before looking at the weights. Hence, the change.

Reply:
Then you can do val weight = train(...)